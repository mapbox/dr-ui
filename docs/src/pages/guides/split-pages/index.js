/*---
title: Split pages
description: This is the first section
splitPages: true
order: 3
layout: page
contentType: API
---*/

import React from 'react';

// Import each markdown file.
import Intro from './sections/intro.md';
import HowTo from './sections/how-to.md';
import Limitations from './sections/limitations.md';
import TagDebugger from './sections/tag-debugger.md';
import PageShell from '../../../components/page-shell';

// Import splitPage function
import splitPages from '@mapbox/batfish/data/split-pages'; // eslint-disable-line

export default class Page extends React.Component {
  render() {
    const { location } = this.props;
    // Override the combines headings in the sidebar
    const headings = splitPages[location.pathname].headings;
    return (
      <PageShell headings={headings} {...this.props}>
        <Intro {...this.props} />
        <HowTo {...this.props} />
        <Limitations {...this.props} />
        <TagDebugger {...this.props} />
      </PageShell>
    );
  }
}
