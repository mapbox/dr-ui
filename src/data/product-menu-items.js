const ProductMenuItems = [
  {
    productCategory: 'Maps',
    icon: 'map',
    products: [
      {
        url: '/mapbox-gl-js/',
        name: 'Mapbox GL JS'
      },
      {
        url: '/ios-sdk/maps/',
        name: 'Maps SDK for iOS'
      },
      {
        url: '/android-docs/maps/',
        name: 'Maps SDK for Android'
      },
      {
        url: '/unity-sdk/',
        name: 'Maps SDK for Unity'
      },
      {
        url: '/help/studio-manual/',
        name: 'Mapbox Studio'
      }
    ]
  },
  {
    productCategory: 'Navigation',
    icon: 'compass',
    products: [
      {
        url: '/ios-sdk/navigation/',
        name: 'Navigation SDK for iOS'
      },
      {
        url: '/android-docs/navigation/overview/',
        name: 'Navigation SDK for Android'
      }
    ]
  },
  {
    productCategory: 'Mapbox Services',
    icon: 'sprocket',
    products: [
      {
        url: '/api-documentation/',
        name: 'Web APIs'
      },
      {
        url: '/android-docs/java/',
        name: 'Java SDK'
      }
    ]
  },
  {
    productCategory: 'More',
    icon: 'plus',
    products: [
      {
        url: '/android-docs/plugins/',
        name: 'Android Plugins'
      },
      {
        url: '/android-docs/core/',
        name: 'Android Core library'
      }
    ]
  }
];

export { ProductMenuItems };
