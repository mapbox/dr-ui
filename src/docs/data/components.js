"use strict";
const React = require("react");
module.exports = [
  {
    name: "BackToTopButton",
    description: null,
    props: null,
    examples: []
  },
  {
    name: "BetaFlag",
    description: null,
    props: {
      tooltipText: {
        type: { name: "string" },
        required: false,
        defaultValue:
          "'This feature is in public beta and is subject to potential changes.'",
        description: <div />
      }
    },
    examples: []
  },
  {
    name: "BookImage",
    description: null,
    props: {
      size: {
        type: { name: "string" },
        required: true,
        defaultValue: undefined,
        description: <div />
      }
    },
    examples: []
  },
  {
    name: "BookletImage",
    description: null,
    props: {
      size: {
        type: { name: "string" },
        required: true,
        defaultValue: undefined,
        description: <div />
      }
    },
    examples: []
  },
  {
    name: "Card",
    description: null,
    props: {
      title: {
        type: { name: "string" },
        required: true,
        defaultValue: undefined,
        description: <div />
      },
      path: {
        type: { name: "string" },
        required: true,
        defaultValue: undefined,
        description: <div />
      },
      description: {
        type: { name: "string" },
        required: true,
        defaultValue: undefined,
        description: <div />
      },
      thumbnail: {
        type: { name: "node" },
        required: false,
        defaultValue: undefined,
        description: <div />
      },
      level: {
        type: { name: "node" },
        required: false,
        defaultValue: undefined,
        description: <div />
      },
      language: {
        type: { name: "string" },
        required: false,
        defaultValue: undefined,
        description: <div />
      }
    },
    examples: []
  },
  {
    name: "CardContainer",
    description: null,
    props: {
      title: {
        type: { name: "string" },
        required: true,
        defaultValue: undefined,
        description: <div />
      },
      path: {
        type: { name: "string" },
        required: true,
        defaultValue: undefined,
        description: <div />
      },
      fullWidthCards: {
        type: { name: "bool" },
        required: true,
        defaultValue: undefined,
        description: <div />
      },
      cards: {
        type: { name: "arrayOf", value: { name: "node" } },
        required: true,
        defaultValue: undefined,
        description: <div />
      }
    },
    examples: []
  },
  {
    name: "CodeToggle",
    description: null,
    props: {
      id: {
        type: { name: "string" },
        required: true,
        defaultValue: undefined,
        description: <div />
      },
      onChange: {
        type: { name: "func" },
        required: true,
        defaultValue: undefined,
        description: <div />
      },
      options: {
        type: {
          name: "arrayOf",
          value: {
            name: "shape",
            value: {
              language: {
                name: "enum",
                value: [
                  { value: "'swift'", computed: false },
                  { value: "'objective-c'", computed: false },
                  { value: "'java'", computed: false },
                  { value: "'kotlin'", computed: false },
                  { value: "'javascript'", computed: false }
                ],
                required: true
              },
              preferredLanguage: { name: "bool", required: true }
            }
          }
        },
        required: true,
        defaultValue: undefined,
        description: <div />
      }
    },
    examples: []
  },
  {
    name: "ContactImage",
    description: null,
    props: {
      size: {
        type: { name: "string" },
        required: true,
        defaultValue: undefined,
        description: <div />
      }
    },
    examples: []
  },
  {
    name: "ExamplesPage",
    description: null,
    props: {
      frontMatter: {
        type: {
          name: "shape",
          value: {
            title: { name: "string", required: true },
            description: { name: "string", required: true }
          }
        },
        required: true,
        defaultValue: undefined,
        description: <div />
      },
      cardContainers: {
        type: { name: "arrayOf", value: { name: "node" } },
        required: true,
        defaultValue: undefined,
        description: <div />
      }
    },
    examples: []
  },
  {
    name: "LevelIndicator",
    description: null,
    props: {
      level: {
        type: {
          name: "enum",
          value: [
            { value: "1", computed: false },
            { value: "2", computed: false },
            { value: "3", computed: false }
          ]
        },
        required: true,
        defaultValue: undefined,
        description: <div />
      }
    },
    examples: []
  },
  {
    name: "NavigationAccordion",
    description: null,
    props: {
      currentPath: {
        type: { name: "string" },
        required: false,
        defaultValue: undefined,
        description: <div />
      },
      contents: {
        type: {
          name: "shape",
          value: {
            firstLevelItems: {
              name: "arrayOf",
              value: {
                name: "shape",
                value: {
                  title: { name: "string", required: true },
                  tag: { name: "node", required: false },
                  path: { name: "string", required: true }
                }
              },
              required: true
            },
            secondLevelItems: {
              name: "arrayOf",
              value: {
                name: "shape",
                value: {
                  title: { name: "string", required: true },
                  path: { name: "string", required: true },
                  thirdLevelItems: {
                    name: "arrayOf",
                    value: {
                      name: "shape",
                      value: {
                        title: { name: "string", required: true },
                        path: { name: "string", required: true }
                      }
                    },
                    required: false
                  }
                }
              },
              required: false
            }
          }
        },
        required: false,
        defaultValue: undefined,
        description: <div />
      },
      onDropdownChange: {
        type: { name: "func" },
        required: false,
        defaultValue: undefined,
        description: <div />
      }
    },
    examples: []
  },
  {
    name: "NavigationDropdown",
    description: null,
    props: {
      currentPath: {
        type: { name: "string" },
        required: true,
        defaultValue: undefined,
        description: <div />
      },
      dropdownOptions: {
        type: {
          name: "arrayOf",
          value: {
            name: "shape",
            value: {
              title: { name: "string", required: true },
              path: { name: "string", required: true }
            }
          }
        },
        required: true,
        defaultValue: undefined,
        description: <div />
      },
      onChange: {
        type: { name: "func" },
        required: false,
        defaultValue: undefined,
        description: <div />
      }
    },
    examples: []
  },
  {
    name: "Note",
    description: null,
    props: {
      title: {
        type: { name: "string" },
        required: false,
        defaultValue: undefined,
        description: <div />
      },
      children: {
        type: { name: "node" },
        required: true,
        defaultValue: undefined,
        description: <div />
      },
      imageComponent: {
        type: { name: "node" },
        required: true,
        defaultValue: undefined,
        description: <div />
      },
      theme: {
        type: { name: "string" },
        required: false,
        defaultValue: undefined,
        description: <div />
      }
    },
    examples: []
  },
  {
    name: "OverviewHeader",
    description: null,
    props: {
      features: {
        type: { name: "arrayOf", value: { name: "string" } },
        required: true,
        defaultValue: undefined,
        description: <div />
      },
      title: {
        type: { name: "string" },
        required: true,
        defaultValue: undefined,
        description: <div />
      },
      beta: {
        type: { name: "bool" },
        required: false,
        defaultValue: "false",
        description: <div />
      },
      image: {
        type: { name: "node" },
        required: true,
        defaultValue: undefined,
        description: <div />
      },
      version: {
        type: { name: "string" },
        required: false,
        defaultValue: undefined,
        description: <div />
      },
      changelogLink: {
        type: { name: "string" },
        required: false,
        defaultValue: undefined,
        description: <div />
      },
      installLink: {
        type: { name: "string" },
        required: false,
        defaultValue: undefined,
        description: <div />
      },
      ghLink: {
        type: { name: "string" },
        required: false,
        defaultValue: undefined,
        description: <div />
      },
      contactLink: {
        type: { name: "string" },
        required: false,
        defaultValue: undefined,
        description: <div />
      }
    },
    examples: []
  },
  {
    name: "PageLayout",
    description: null,
    props: {
      sidebarContent: {
        type: { name: "node" },
        required: true,
        defaultValue: undefined,
        description: <div />
      },
      sidebarTitle: {
        type: { name: "union", value: [{ name: "node" }, { name: "string" }] },
        required: false,
        defaultValue: undefined,
        description: <div />
      },
      sidebarTheme: {
        type: { name: "string" },
        required: false,
        defaultValue: "'bg-gray-faint'",
        description: <div />
      },
      sidebarContentStickyTop: {
        type: { name: "number" },
        required: true,
        defaultValue: undefined,
        description: <div />
      },
      sidebarContentStickyTopNarrow: {
        type: { name: "number" },
        required: true,
        defaultValue: undefined,
        description: <div />
      },
      sidebarStackedOnNarrowScreens: {
        type: { name: "bool" },
        required: false,
        defaultValue: "false",
        description: <div />
      },
      sideBarColSize: {
        type: { name: "number" },
        required: false,
        defaultValue: undefined,
        description: <div />
      },
      interactiveClass: {
        type: { name: "string" },
        required: false,
        defaultValue: undefined,
        description: <div />
      },
      children: {
        type: { name: "node" },
        required: true,
        defaultValue: undefined,
        description: <div />
      }
    },
    examples: []
  },
  {
    name: "ProductMenu",
    description: null,
    props: {
      productName: {
        type: { name: "union", value: [{ name: "string" }, { name: "node" }] },
        required: true,
        defaultValue: undefined,
        description: <div />
      },
      beta: {
        type: { name: "bool" },
        required: false,
        defaultValue: "false",
        description: <div />
      },
      lightText: {
        type: { name: "bool" },
        required: false,
        defaultValue: "false",
        description: <div />
      },
      homePage: {
        type: { name: "string" },
        required: true,
        defaultValue: undefined,
        description: <div />
      }
    },
    examples: []
  },
  {
    name: "Search",
    description: null,
    props: {
      placeholder: {
        type: { name: "string" },
        required: false,
        defaultValue: "'Search docs.mapbox.com'",
        description: <div />
      },
      narrow: {
        type: { name: "bool" },
        required: false,
        defaultValue: undefined,
        description: <div />
      },
      background: {
        type: {
          name: "enum",
          value: [
            { value: "'light'", computed: false },
            { value: "'dark'", computed: false }
          ]
        },
        required: false,
        defaultValue: "'light'",
        description: <div />
      },
      inputId: {
        type: { name: "string" },
        required: false,
        defaultValue: "'docs-search'",
        description: <div />
      },
      disableModal: {
        type: { name: "bool" },
        required: false,
        defaultValue: undefined,
        description: <div />
      }
    },
    examples: []
  },
  {
    name: "SectionedNavigation",
    description: null,
    props: {
      sections: {
        type: { name: "arrayOf", value: { name: "object" } },
        required: true,
        defaultValue: undefined,
        description: <div />
      },
      title: {
        type: { name: "string" },
        required: false,
        defaultValue: undefined,
        description: <div />
      },
      includeCount: {
        type: { name: "bool" },
        required: false,
        defaultValue: "true",
        description: <div />
      },
      includeFilterBar: {
        type: { name: "bool" },
        required: false,
        defaultValue: "false",
        description: <div />
      },
      hideSubItems: {
        type: { name: "bool" },
        required: false,
        defaultValue: "false",
        description: <div />
      }
    },
    examples: []
  },
  {
    name: "ToggleableCodeBlock",
    description: null,
    props: {
      codeSnippet: {
        type: {
          name: "arrayOf",
          value: {
            name: "shape",
            value: {
              language: {
                name: "enum",
                value: [
                  { value: "'swift'", computed: false },
                  { value: "'objective-c'", computed: false },
                  { value: "'java'", computed: false },
                  { value: "'kotlin'", computed: false },
                  { value: "'javascript'", computed: false }
                ],
                required: false
              },
              rawCode: { name: "string", required: true },
              highlightedCode: { name: "string", required: true },
              preferredLanguage: { name: "bool", required: true }
            }
          }
        },
        required: true,
        defaultValue: undefined,
        description: <div />
      }
    },
    examples: []
  },
  {
    name: "TopbarSticker",
    description: null,
    props: {
      unStickWidth: {
        type: { name: "number" },
        required: false,
        defaultValue: undefined,
        description: <div />
      },
      children: {
        type: { name: "node" },
        required: true,
        defaultValue: undefined,
        description: <div />
      }
    },
    examples: []
  },
  {
    name: "TroubleshootImage",
    description: null,
    props: {
      size: {
        type: { name: "string" },
        required: true,
        defaultValue: undefined,
        description: <div />
      }
    },
    examples: []
  },
  {
    name: "WarningImage",
    description: null,
    props: {
      size: {
        type: { name: "string" },
        required: true,
        defaultValue: undefined,
        description: <div />
      },
      color: {
        type: { name: "string" },
        required: false,
        defaultValue: undefined,
        description: <div />
      }
    },
    examples: []
  }
];
