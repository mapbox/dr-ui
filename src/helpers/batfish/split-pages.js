// builds all top level sections and their split pages
// this dataset is used by:
// API: IntroToc, IntroServicesList, and powers the headings for each section page
// Studio manual
// Accounts

const fs = require('fs');
const remark = require('remark-parse');
const unified = require('unified');
const visit = require('unist-util-visit');
const frontmatter = require('remark-frontmatter');
const GithubSlugger = require('github-slugger');

const slugger = new GithubSlugger();

const sortBy = (key) => {
  return (a, b) => (a[key] > b[key] ? 1 : b[key] > a[key] ? -1 : 0);
};

function buildSplitPages(data) {
  // get all pages with `splitPages: true`
  const mainSplitPages = data.pages.filter((f) => f.frontMatter.splitPages);
  // find each sub pages to the top level page
  return mainSplitPages.reduce((obj, page) => {
    const splitPages = findSplitPages(data, page);
    obj[page.path] = {
      ...page,
      pages: splitPages,
      headings: combine(splitPages, 'headings')
    };
    return obj;
  }, {});
}

function findSplitPages(data, page) {
  return data.pages
    .filter((f) => f.path.startsWith(page.path) && f.frontMatter.splitPage)
    .map((page) => ({
      ...page,
      headings: getHeadings(page),
      slug: slugger.slug(page.frontMatter.title),
      order: page.frontMatter.order // pull out order for easy sorting
    }))
    .sort(sortBy('order'));
}

function combine(pages, field) {
  const combinedArray = pages.reduce((arr, page) => {
    if (page[field]) arr = arr.concat(page[field]);
    return arr;
  }, []);
  return uniq(combinedArray);
}

function getHeadings(page) {
  let headings = [];
  const content = fs.readFileSync(page.filePath, 'utf-8');
  const ast = unified().use(remark).use(frontmatter, ['yaml']).parse(content);
  visit(ast, 'heading', (node) => {
    // only capture h2 and h3
    if (node.depth > 3) return;
    const label = node.children.reduce((str, n) => {
      if (n.type === 'text' && n.value) str += n.value;
      return str;
    }, '');
    const firstNode = node.children[0].value;

    headings.push({
      level: node.depth,
      text: label,
      slug: slugger.slug(firstNode),
      ...(page.frontMatter.tag &&
        node.depth === 2 && { tag: page.frontMatter.tag }) // only show tag for first item
    });
  });
  slugger.reset(); // reset slugger
  return headings;
}

// return a unique array
function uniq(arr) {
  if (!arr) return [];
  return arr.filter((elem, index) => {
    return arr.indexOf(elem) === index;
  });
}

module.exports = {
  buildSplitPages
};
