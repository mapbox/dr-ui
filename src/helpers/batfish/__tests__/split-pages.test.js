import { buildSplitPages } from '../split-pages';
import data from './fixtures/data.json';

it('buildSplitPages', () => {
  expect(buildSplitPages(data)).toMatchSnapshot();
});
