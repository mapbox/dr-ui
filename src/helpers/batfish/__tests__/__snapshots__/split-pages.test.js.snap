// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildSplitPages 1`] = `
Object {
  "/dr-ui/guides/split-pages/": Object {
    "filePath": "./docs/src/pages/guides/split-pages/index.js",
    "frontMatter": Object {
      "contentType": "API",
      "description": "This is the first section",
      "layout": "page",
      "order": 3,
      "splitPages": true,
      "title": "Split pages",
    },
    "headings": Array [
      Object {
        "level": 2,
        "slug": "how-to-use-split-pages",
        "text": "How to use split pages",
      },
      Object {
        "level": 3,
        "slug": "1-create-the-main-page",
        "text": "1. Create the main page",
      },
      Object {
        "level": 3,
        "slug": "2-create-the-split-pages",
        "text": "2. Create the split pages",
      },
      Object {
        "level": 3,
        "slug": "3-update-the-batfish-configuration",
        "text": "3. Update the Batfish configuration",
      },
      Object {
        "level": 3,
        "slug": "4-create-redirects",
        "text": "4. Create redirects",
      },
      Object {
        "level": 2,
        "slug": "limitations",
        "text": "Limitations",
      },
      Object {
        "level": 2,
        "slug": "tag-debugger",
        "tag": "beta",
        "text": "Tag debugger",
      },
    ],
    "pages": Array [
      Object {
        "filePath": "./docs/src/pages/guides/split-pages/sections/intro.md",
        "frontMatter": Object {
          "contentType": "guide",
          "description": "An introduction to use split pages.",
          "hideFeedback": true,
          "order": 1,
          "products": Array [
            "Documentation",
          ],
          "splitPage": true,
          "title": "Introduction to split pages",
        },
        "headings": Array [],
        "order": 1,
        "path": "/dr-ui/guides/split-pages/sections/intro/",
        "slug": "introduction-to-split-pages",
      },
      Object {
        "filePath": "./docs/src/pages/guides/split-pages/sections/how-to.md",
        "frontMatter": Object {
          "contentType": "guide",
          "description": "An introduction to use split pages.",
          "hideFeedback": true,
          "order": 2,
          "prependJs": Array [
            "import MainPage from '!raw-loader!../index.js'; //eslint-disable-line",
            "import SubOne from '!raw-loader!./intro.md'; //eslint-disable-line",
            "import SplitPageShell from '!raw-loader!../../../../components/split-page-shell.js'; //eslint-disable-line",
            "import CodeSnippet from '../../../../../../src/components/code-snippet/code-snippet';",
            "import { highlightJsx } from '../../../../../../src/components/highlight/jsx';",
            "import { highlightHtml } from '../../../../../../src/components/highlight/html';",
          ],
          "products": Array [
            "Mapbox fundamentals",
          ],
          "splitPage": true,
          "title": "How to use spit pages",
        },
        "headings": Array [
          Object {
            "level": 2,
            "slug": "how-to-use-split-pages",
            "text": "How to use split pages",
          },
          Object {
            "level": 3,
            "slug": "1-create-the-main-page",
            "text": "1. Create the main page",
          },
          Object {
            "level": 3,
            "slug": "2-create-the-split-pages",
            "text": "2. Create the split pages",
          },
          Object {
            "level": 3,
            "slug": "3-update-the-batfish-configuration",
            "text": "3. Update the Batfish configuration",
          },
          Object {
            "level": 3,
            "slug": "4-create-redirects",
            "text": "4. Create redirects",
          },
        ],
        "order": 2,
        "path": "/dr-ui/guides/split-pages/sections/how-to/",
        "slug": "how-to-use-spit-pages",
      },
      Object {
        "filePath": "./docs/src/pages/guides/split-pages/sections/limitations.md",
        "frontMatter": Object {
          "contentType": "guide",
          "description": "Limitations to the split page pattern.",
          "hideFeedback": true,
          "order": 3,
          "products": Array [
            "Mapbox fundamentals",
          ],
          "splitPage": true,
          "title": "Limitations",
        },
        "headings": Array [
          Object {
            "level": 2,
            "slug": "limitations",
            "text": "Limitations",
          },
        ],
        "order": 3,
        "path": "/dr-ui/guides/split-pages/sections/limitations/",
        "slug": "limitations",
      },
      Object {
        "filePath": "./docs/src/pages/guides/split-pages/sections/tag-debugger.md",
        "frontMatter": Object {
          "contentType": "guide",
          "description": "Limitations to the split page pattern.",
          "hideFeedback": true,
          "order": 4,
          "products": Array [
            "Mapbox fundamentals",
          ],
          "splitPage": true,
          "tag": "beta",
          "title": "Tag debugger",
        },
        "headings": Array [
          Object {
            "level": 2,
            "slug": "tag-debugger",
            "tag": "beta",
            "text": "Tag debugger",
          },
        ],
        "order": 4,
        "path": "/dr-ui/guides/split-pages/sections/tag-debugger/",
        "slug": "tag-debugger",
      },
    ],
    "path": "/dr-ui/guides/split-pages/",
  },
}
`;
